modules/r2dbc/src/main/java/org/testcontainers/r2dbc/ConnectionPublisher.java

```diff
@@ -0,0 +1,152 @@
+package org.testcontainers.r2dbc;
+
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionFactory;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+class ConnectionPublisher implements Publisher<Connection> {
+
+    private final Supplier<CompletableFuture<ConnectionFactory>> futureSupplier;
+
+    ConnectionPublisher(Supplier<CompletableFuture<ConnectionFactory>> futureSupplier) {
+        this.futureSupplier = futureSupplier;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super Connection> actual) {
+        actual.onSubscribe(new StateMachineSubscription(actual));
+    }
+
+    /**
+     * Design notes:
+     * - ConnectionPublisher is Mono-like (0..1), the request amount is ignored
+     * - given the testing nature, the performance requirements are less strict
+     * - "synchronized" is used to avoid races
+     * - Reactive Streams spec violations are not checked (e.g. non-positive request)
+     */
+    private class StateMachineSubscription implements Subscription {
+
+        private final Subscriber<? super Connection> actual;
+
+        Subscription subscriptionState;
+
+        StateMachineSubscription(Subscriber<? super Connection> actual) {
+            this.actual = actual;
+            subscriptionState = new WaitRequestSubscriptionState();
+        }
+
+        @Override
+        public synchronized void request(long n) {
+            subscriptionState.request(n);
+        }
+
+        @Override
+        public synchronized void cancel() {
+            subscriptionState.cancel();
+        }
+
+        synchronized void transitionTo(SubscriptionState newState) {
+            subscriptionState = newState;
+            newState.enter();
+        }
+
+        abstract class SubscriptionState implements Subscription {
```

Comment:
that pattern was a bit harder for me to follow because of the habit of seeing `Subscription` as self-sufficient and thus aggressively guarded, but after review it looks correct. I feel it wouldn't be too hard to collapse the various `SubscriptionState` into the `StateMachineSubscription` itself and deal with int/enum based states for transitions, but eh as long as it is correct...

Reply:
Thanks for looking at it! FYI I just fixed a race reported by @OlegDokuka:
https://github.com/testcontainers/testcontainers-java/pull/2545/commits/d3409df8e09c28f0f3531d9f2918e55ebd37eebb